core/domain/email_manager_test.py

```diff
@@ -88,6 +89,131 @@ def test_sender_id_validation(self):
+            self.put_json('%s/%s' % (
+                feconf.EXPLORATION_RIGHTS_PREFIX, self.exploration.id), {
+                    'version': self.exploration.version,
+                    'new_member_username': self.NEW_USER_USERNAME,
+                    'new_member_role': 'editor',
+                }, csrf_token=csrf_token)
+
+            messages = self.mail_stub.get_sent_messages(to=self.NEW_USER_EMAIL)
+            self.assertEqual(len(messages), 1)
+
+    def test_email_is_not_sent_if_user_do_not_wish_to(self):
+        user_services.update_email_preferences(self.new_user_id, True, False)
+
+        can_send_emails_ctx = self.swap(
+            feconf, 'CAN_SEND_EMAILS_TO_USERS', True)
+        logged_errors = []
+
+        def _log_error_for_tests(error_message):
+            logged_errors.append(error_message)
+
+        log_new_error_counter = test_utils.CallCounter(_log_error_for_tests)
+        log_new_error_ctx = self.swap(
+            email_manager, 'log_new_error', log_new_error_counter)
+
+        with can_send_emails_ctx, log_new_error_ctx:
+            email_manager.send_exploration_membership_email(
+                self.editor_id, self.new_user_id, 'owner',
+                self.exploration.id, self.exploration.title)
+
+            self.assertEqual(log_new_error_counter.times_called, 1)
+
+    def test_email_correct_membership_email_are_sent(self):
+        can_send_emails_ctx = self.swap(
+            feconf, 'CAN_SEND_EMAILS_TO_USERS', True)
+
+        email_subject = '%s added you to one of their explorations on Oppia.org'
+        with can_send_emails_ctx:
+            email_manager.send_exploration_membership_email(
+                self.editor_id, self.new_user_id, 'viewer',
+                self.exploration.id, self.exploration.title)
+
+            messages = self.mail_stub.get_sent_messages(to=self.NEW_USER_EMAIL)
+            self.assertEqual(len(messages), 1)
+
+            all_models = email_models.SentEmailModel.get_all().fetch()
+            self.assertEqual(len(all_models), 1)
+
+            sent_email_model = all_models[0]
+
+            # Check that email details are correct.
+            self.assertEqual(
+                sent_email_model.recipient_id,
+                self.new_user_id)
+            self.assertEqual(
+                sent_email_model.recipient_email, self.NEW_USER_EMAIL)
+            self.assertEqual(
+                sent_email_model.sender_id, feconf.SYSTEM_COMMITTER_ID)
+            self.assertEqual(
+                sent_email_model.sender_email,
+                'Site Admin <%s>' % feconf.SYSTEM_EMAIL_ADDRESS)
+            self.assertEqual(
+                sent_email_model.intent,
+                feconf.EMAIL_INTENT_EDITOR_ROLE_NOTIFICATION)
+            self.assertEqual(
+                sent_email_model.subject,
+                email_subject % self.EDITOR_USERNAME)
+
+    def test_correct_rights_are_written_in_email_body(self):
```

Comment:
So basically there should be 3 tests, each for checking email body for each of the 3 rights ?